voip_utils/sip.py

```diff
@@ -178,99 +310,193 @@ def datagram_received(self, data: bytes, addr):
+                # TODO: Verify that the call / sequence IDs match our outgoing INVITE
+                _LOGGER.debug("Received response [%s]", message)
+                is_ok = smsg.code == "200" and smsg.reason == "OK"
+                if not is_ok:
+                    _LOGGER.debug("Received non-OK response [%s]", message)
+                    return
+
+                _LOGGER.debug("Got OK message")
+                if self.transport is None:
+                    _LOGGER.warning("No transport for exchanging SIP message")
+                    return
+                rtp_info = get_rtp_info(smsg.body)
+                remote_rtp_ip = rtp_info.rtp_ip
+                remote_rtp_port = rtp_info.rtp_port
+                opus_payload_type = rtp_info.payload_type
+                caller_endpoint = None
+                if smsg.headers.get("to") is not None:
+                    caller_endpoint = SipEndpoint(smsg.headers.get("to", ""))
+                else:
+                    caller_endpoint = get_sip_endpoint(caller_ip, port=caller_sip_port)
+                # The From header should give us the URI used for sending SIP messages to the device
+                local_endpoint = None
+                if smsg.headers.get("from") is not None:
+                    local_endpoint = SipEndpoint(smsg.headers.get("from", ""))
+                else:
+                    local_endpoint = get_sip_endpoint(caller_ip, port=caller_sip_port)
+
+                _LOGGER.debug("Outgoing call to endpoint=%s", caller_endpoint)
+                ack_lines = [
+                    f"ACK {caller_endpoint.uri} SIP/2.0",
+                    f"Via: SIP/2.0/UDP {local_endpoint.host}:{local_endpoint.port}",
+                    f"From: {local_endpoint.sip_header}",
+                    f"To: {smsg.headers['to']}",
+                    f"Call-ID: {smsg.headers['call-id']}",
+                    "CSeq: 50 ACK",
+                    f"User-Agent: {VOIP_UTILS_AGENT} 1.0",
+                    "Content-Length: 0",
+                ]
+                ack_text = _CRLF.join(ack_lines) + _CRLF
+                ack_bytes = ack_text.encode("utf-8")
+                self.transport.sendto(ack_bytes, (caller_ip, caller_sip_port))
+
+                # The call been answered, proceed with desired action here
+                local_rtp_port = self._get_call_rtp_port(smsg.headers["call-id"])
+                self.on_call(
+                    CallInfo(
+                        caller_endpoint=caller_endpoint,
+                        caller_rtp_port=remote_rtp_port,
+                        server_ip=remote_rtp_ip,
+                        headers=smsg.headers,
+                        opus_payload_type=opus_payload_type,  # Should probably update this to eventually support more codecs
+                        local_rtp_ip=local_endpoint.host,
+                        local_rtp_port=local_rtp_port,
+                    )
+                )
+            elif method == "bye":
```

Comment:
Should this cleanup the call / call an abstract method to indicate the call is over? We would have to update the state in HA

Reply:
It was already calling a method to clean up this class' call dictionary, but I added an abstract on_hangup method as well.